livecheck: progress bar for JSON output

nandahkrishna opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This PR adds a progress bar for livecheck's JSON output. We often use the `--json` flag when we have too many formulae to check and would like to redirect the livecheck output to a file, maybe to use later in a script or to compare livecheck outputs after making changes to livecheckables.

Currently, no feedback or progress information is provided, and we're left wondering how long it's going to take until livecheck is done. The addition of the progress bar keeps us updated on the status and helps us anticipate how much longer the process would take. I used the `ruby-progressbar` gem as I was unable to find any other command or module which had a progress bar (`curl` uses its own `--progress-bar` option), so do let me know if I should be doing it some other way.

This progress bar is temporary – it stays on-screen until livecheck is done preparing the JSON output, following which it is erased (for a TTY stdout). This 'feature' required the addition of some additional `SPECIAL_CODES` to the `Tty` module (to clear a line and move up a line, I added a few more to make sure we have a 'complete' set).

I think I've covered all the details here. Looking forward to your comments/suggestions!

CC @samford

samford commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -66,6 +67,20 @@ def livecheck_formulae(formulae_to_check, args)
+          formulae_to_check.count
+        else
+          formulae_to_check.length
+        end
+        ohai "Generating JSON"
+        progress = ProgressBar.create(
+          total:         total_formulae,
+          progress_mark: "#",
+          format:        "%b %c/%C",
+        )
+      end
```
```suggestion
        end

        ohai "Generating JSON"
        progress = ProgressBar.create(
          total:  total_formulae,
          format: " %t: |%b%i| %c/%C ",
        )
      end

```

What do you think about this format? It resolves a few issues that I was experiencing when testing it:

* Added an extra space at the start, so the first character is more visible (and adds a space at the end to mimic the leading space).
* Added an indicator of the start/end of the progress bar (`|`).
* Filled the incomplete space in the progress bar, so the finished/total numbers are aligned to the right.

nandahkrishna replied:
That's actually what I had before creating the PR (like the bars in my scripts), haha. The reason I changed it was to maintain some consistency and make it look similar to the `curl` download progress bar, which seems to take up `0.8 * Tty.width` (72 #s on completion and my terminal width is 90).

I like this format better, so I don't mind changing it.

samford commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -66,6 +67,20 @@ def livecheck_formulae(formulae_to_check, args)
 
       has_a_newer_upstream_version = false
+
+      if args.json?
```
```suggestion
      if args.json? && !args.quiet?
```

nandahkrishna replied:
I had a doubt about this, the help message for `--quiet` says that it should be used to suppress warnings. Would it be okay to use it for this purpose (be non-verbose)?

samford replied:
Hmm, that's a good question. I wouldn't want to misuse it, so I'll ask internally and see what the general consensus is.

If this isn't appropriate, then our options would seemingly be to either put the progress bar behind a flag (`--progress`) or add a flag to avoid the progress bar (`--no-progress`). Anyone running the command would specifically have to know about the `no-progress` flag to avoid the progress output when running the command in scripts, so I would probably err on the side of making the progress bar opt-in via a `--progress` flag (if we have to go down this route).

nandahkrishna replied:
I've decided (for now) to implement the `--progress` flag and the `--output-file=` argument. I've finished it locally but the command completions need to be updated, so I'll push after I'm done with those.

MikeMcQuaid replied:
> Would it be okay to use it for this purpose (be non-verbose)

Yup I think so 👍🏻

samford commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -136,6 +151,7 @@ def livecheck_formulae(formulae_to_check, args)
         has_a_newer_upstream_version ||= true
 
         if args.json?
+          progress.increment
```
```suggestion
          progress.increment if progress
```

samford commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -157,7 +174,11 @@ def livecheck_formulae(formulae_to_check, args)
-      puts JSON.generate(formulae_checked.compact) if args.json?
+      return unless args.json?
+
+      progress.finish
+      print "#{Tty.up}#{Tty.erase_line}" * 2
```
```suggestion
      if progress
        progress.finish
        print "#{Tty.up}#{Tty.erase_line}" * 2
      end

```

issyl0 commented:
I was going to ask how long a livecheck run usually takes, but I decided to try it myself:

```
issyl0 on ophiuchus in /usr/local/Homebrew on livecheck-json-progress via 💎 v2.6.3
➜ brew livecheck --all --json
==> Generating JSON
############## 398/5530
```

And it's still going. I can see why this might be good to have!

samford reviewed (changes requested):
This is useful for when you're doing a run where you're not redirecting the output to a file and does deal with the issue where we have to wait in silence until livecheck spits out all the JSON at the end. I added a suggestion that provides a different progress bar format which may be a little more pleasant in some respects.

Unfortunately, the progress bar doesn't seem to be compatible with redirecting the output to a file. I do something like `brew livecheck --json --tap homebrew/core > some_directory/filename.json` very frequently and this ends up producing a file that isn't valid JSON since the progress bar output appears at the beginning of the file.

One way of dealing with this would be to add a CLI option like `--output-file` (or something like that) where livecheck would write the JSON (making sure we can write there before doing the work, of course). I imagine we would only allow this option when the `--json` flag is also present (until/unless someone comes along asking to do this for the normal output). This would allow us to have the best of both worlds, where we could have the progress bar while also outputting the JSON to a file. This particular setup is exactly what I'm looking for.

The other issue is that the progress bar output would also appear in the output if you were to run the command in a script and capture the output (as happens in the `utils/livecheck_formula.rb` file that `brew bump` uses), so this would also result in invalid JSON. One way of dealing with this would be to avoid printing the progress bar when the `--quiet` flag is present and I added suggestions for this.

samford commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -146,6 +162,7 @@ def livecheck_formulae(formulae_to_check, args)
         Homebrew.failed = true
 
         if args.json?
+          progress.increment
```
```suggestion
          progress.increment if progress
```

I missed this one the first time around.

MikeMcQuaid commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -1,6 +1,7 @@
 # frozen_string_literal: true
 
 require "livecheck/strategy"
+require "ruby-progressbar"
```
Is this a gem that would be vendored? How many dependencies does it have? As @samford mentioned: does it support TTY handling? If not: we should handle this ourselves.

nandahkrishna replied:
We already have it vendored I think. Known for not having dependencies based on what I saw.

MikeMcQuaid replied:
> Known for not having dependencies based on what I saw.

Cool!



> We already have it vendored I think.

It's vendored but `gitignored` so https://github.com/Homebrew/brew/blob/master/.gitignore#L133 would need removed.



> As for TTY handling, I erase the progress bar using our own `Tty` module.

I was thinking it should be disabled if it's not a TTY e.g. `unless $stdout.tty?`.

nandahkrishna replied:
Alright, I'll go ahead and just not have the progress bar unless it's a TTY 👍🏼

nandahkrishna commented:
With the latest push:
* Fixed a situation in which the head downloader output wouldn't be suppressed for a non-HEAD-only formula (but the installed version is `head`) – replaced `formula.head.downloader.shutup! if formula.head?` with `formula.head&.downloader&.shutup!`.
* Added an `--output-file` option.
* Added the `$stdout.tty?` condition to print the progress bar.
* Using `--quiet` suppresses the progress bar, and the help string for the flag includes this information.
* The `ruby-progressbar` dependency is now vendored (removed from `.gitignore`).
* Updated the command completions and manpage.

I think I've covered most (if not all) changes made. Let me know if anything else should be changed or improved. Thanks!

samford commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -66,6 +67,20 @@ def livecheck_formulae(formulae_to_check, args)
+        end
+        ohai "Generating JSON"
+        progress = ProgressBar.create(
+          total:         total_formulae,
+          progress_mark: "#",
+          format:        " %t: |%B| %c/%C ",
```
```suggestion
          total:          total_formulae,
          progress_mark:  "#",
          remainder_mark: ".",
          format:         " %t: [%B] %c/%C ",
```

I was updating a server yesterday and found some inspiration. This version takes some cues from Ubuntu's progress bar, with square brackets at the start/end and periods as remainder marks:

```
 Progress: [###..........................................................] 1/17
```

I think this is the last time I'll tweak this. What do you think about this format?

nandahkrishna replied:
Looks good!

samford reviewed (commented):
I can only comment on the changes within livecheck (and I'll rely on someone else to say whether the other parts are correct) but I think this should be good from my perspective after these final comments are resolved.

I did a run with the progress bar and `--output-file=` and it's quite nice 👍

nandahkrishna commented on Library/Homebrew/dev-cmd/livecheck.rb:
```diff
@@ -48,6 +53,8 @@ def livecheck
       puts ENV["HOMEBREW_LIVECHECK_WATCHLIST"] if ENV["HOMEBREW_LIVECHECK_WATCHLIST"].present?
     end
 
+    File.open(args.output_file, "w").close if args.output_file.present?
```
Ah that's something that didn't cross my mind. Having explicit error messages is good so I'll make change 👍🏼

samford commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -66,6 +67,20 @@ def livecheck_formulae(formulae_to_check, args)
+        else
+          formulae_to_check.length
+        end
+        ohai "Generating JSON"
```
```suggestion
        ohai "Running checks"
```

One last thing, I think. We technically don’t generate the JSON until the end, so “Running checks” is more accurate and also makes it clear what the progress bar is tracking.

nandahkrishna replied:
Haha yes, I always felt the message was not exactly apt. Making the change 👍🏼

dtrodrigues commented:
This isn't strictly needed given the `--output-file` option, but what about also/instead having the progress bar write to stderr instead of stdout which would allow you to do `brew livecheck --json > output.txt` while still getting the progress bar on the screen and not polluting the output file?

dtrodrigues commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -66,6 +67,21 @@ def livecheck_formulae(formulae_to_check, args)
+          total:          total_formulae,
+          progress_mark:  "#",
+          remainder_mark: ".",
+          format:         " %t: [%B] %c/%C ",
```
```suggestion
          format:         " %t: [%B] %c/%C ",
          output:         $stderr,
```

nandahkrishna replied:
Although this solves our problem, would it be _okay_ to write non-error output to `stderr` by default?

dtrodrigues replied:
Yeah, `stderr` is often for diagnostic information, not just errors -- https://unix.stackexchange.com/a/331808.
The curl progress bar is written to `stderr`, for instance.

nandahkrishna replied:
TIL, thank you! The link was very helpful.

dtrodrigues commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -66,6 +67,21 @@ def livecheck_formulae(formulae_to_check, args)
 
       has_a_newer_upstream_version = false
+
+      if args.json? && !args.quiet? && $stdout.tty?
```
```suggestion
      if args.json? && !args.quiet?
```

nandahkrishna commented:
Now that we have a solution to the redirection problem by writing the progress bar to `stderr`, do we still need `--output-file?`

dtrodrigues commented:
I personally think there would be no need for `--output-file` and it would make the codebase simpler without it, but there may be a use case I'm not thinking of.

MikeMcQuaid commented:
> I personally think there would be no need for `--output-file` and it would make the codebase simpler without it, but there may be a use case I'm not thinking of.

Agreed. I don't think we have any other commands that take it but many that are designed to work differently if `brew $COMMAND > file` (which should be handled by the TTY check here).

samford commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -66,6 +67,22 @@ def livecheck_formulae(formulae_to_check, args)
+          formulae_to_check.count
+        else
+          formulae_to_check.length
+        end
```
```suggestion
        end

```

A little whitespace for style purposes (to make this more readable).

samford commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -66,6 +67,22 @@ def livecheck_formulae(formulae_to_check, args)
+          format:         " %t: [%B] %c/%C ",
+          output:         $stderr,
+        )
+      end
```
```suggestion
      end

```

I think I suggested this in a previous review but maybe we lost the change along the way. This blank line simply separates this `if` block from the code after it, for the sake of readability.

samford commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -157,7 +176,14 @@ def livecheck_formulae(formulae_to_check, args)
+
+      if progress
+        progress.finish
+        print "#{Tty.up}#{Tty.erase_line}" * 2
```
```suggestion
        $stderr.print "#{Tty.up}#{Tty.erase_line}" * 2
```

This should print to $stderr now, correct?

nandahkrishna replied:
It works even otherwise which is why I didn't change it, but I'll make the change now.

nandahkrishna replied:
Ah, it seemed to work without it normally but I realise I didn't check the redirection part. I've made the change now.

samford commented on Library/Homebrew/utils/tty.rb:
```diff
@@ -66,6 +75,16 @@ def reset_escape_sequence!
 
+  SPECIAL_CODES.each do |name, code|
+    define_singleton_method(name) do
+      if $stdout.tty?
```
I'm not sure of the best way to update this but it needs to be modified in some way. Right now, the progress bar isn't cleared at the end of the run when `$stdout` is redirected to a file (e.g., `brew livecheck --json > test.json`).

nandahkrishna replied:
Despite `stderr.tty?` being true, I don't see coloured output (the `Running checks` line), is there something I'm missing?

samford replied:
My tentative solution for this was to replace this line with `if $stdout.tty? || $stderr.tty?`. Fixing the color issue involves a similar change in `#color?`, where we could change `$stdout.tty?` to `$stdout.tty? || $stderr.tty?`.

That said, I'm not sure if this is the most appropriate way to handle this. It works for our particular purposes but it would be worth making sure that this won't negatively impact some other usage of `utils/tty` in Homebrew.

nandahkrishna replied:
Got it 👍🏼 I'll dig a little deeper and see what's going on. Thanks Sam.

nandahkrishna replied:
I went through the usages after identifying them using `grep -R`, and I don't think there's going to be any untoward impact. So I've gone ahead and made this change.

nandahkrishna commented:
With the current `utils/tty` implementation, redirection of `stdout` is handled (we make sure we check if `stdout` is a TTY in order to print the special characters and colours).

However the problems are:
* Redirection of only `stderr` to a file: if there are coloured outputs in stderr, the special characters feature in the file. For example, without a watchlist file, `brew livecheck 2> test.txt` results in `text.txt` containing colour-code characters.
* Redirection of only `stdout`: `stderr` loses colour output.

I tried adding an extra parameter to the methods in `utils/tty`, which would be `$stdout` by default – so we could check for only one of `stderr` or `stdout` asper the situation. However, this would require a large number of changes which I believe should be another PR.

Any thoughts on how we should move forward? To me, it seems like we should revert to the `--output-file` setup for now, get the `tty` changes done (if they can/should be) and then fix this again.

samford commented:
> Any thoughts on how we should move forward?

The general consensus is that we shouldn't have an `--output-file` flag and should just rely on output redirection, so we shouldn't reintroduce that. It seems like the options are:

1. Go back to printing the progress bar on `stdout` for the time being and temporarily live with the progress bar not displaying when `stdout` is redirected to a file (e.g., `brew livecheck --json > file.json`).
1. Continue printing the progress bar on `stderr` but don't modify the `$stdout` conditions in `utils/tty.rb`. The progress bar wouldn't have color and wouldn't be cleared when `stdout` is redirected to a file but this would at least allow us to have some form of the progress bar in this situation for the interim time.
1. Handle the `Tty` changes in a separate PR and simply return to this PR (rebasing and merging) after they're done.

I think how we proceed here depends on any `Tty` changes and how much work would be involved in a PR for that. I'll defer to @MikeMcQuaid on whether your `Tty` suggestion (i.e., an `@output` instance variable) is the right way to go or if there's another option.

nandahkrishna commented:
I'll go ahead with option 3 and get the `Tty` PR done, so we can make some progress (no pun intended, haha). I'll return to this once the other issues have been sorted out.

MikeMcQuaid reviewed (approved):
Fine with me once @samford is happy. Thanks @nandahkrishna!

samford commented:
I tested this branch locally by rebasing on the latest main branch and making some `Tty`-related changes (with respect to #8624). With the changes in place, `brew livecheck --json > some_file.json` properly displays color in the `$stderr` output and clears the output (i.e., "==> Running checks" and the progress bar) when the run finishes. I did a full JSON run across the homebrew/core formulae with no problems.

I think this should be good after the related changes have been incorporated into this PR.

nandahkrishna commented:
@samford: I've rebased and incorporated the changes, let me know if this is good to merge, thanks.

samford reviewed (approved):
Looks good to me now. Thanks, Nanda!
